rfcs/shared/build-system/package-public-api.md

```diff
@@ -0,0 +1,237 @@
+# RFC: Package public API accessors
+
+---
+
+_@hotell_
+
+## Summary
+
+We wanna control which parts of API surface of our packages are being exposed to public.
+
+## Background
+
+During v9 pre-release phase we ran into various situation where we didn't want to propagate APIs as ready to go, thus we established `_unstable` identifiers naming suffix to mark those as unstable. To complement this approach we also introduced `*/unstable` deep imports for API's that are not STABLE. This works to some extent but fails short as its a manual work that is not checked by automation. This led to scenarios that we unintentionally exposed INTERNAL apis to public which led to Breaking changes which we wanna avoid at all costs as we follow proper semver.
+
+## Problem statement
+
+We expose all APIs of our packages as a PUBLIC from barrel files (`src/index.ts`) to consumers, which introduces problems as we can expose apis that were supposed to be used only internally within our repo or only be available in particular pre-release stage (ALPHA/BETA versions).
+
+## Detailed Design or Proposal
+
+We wanna provide style guide/common approach how to annotate package APIs with proper ACCESS modifiers which will be processed/verified by tooling automation.
+
+Note that our approach will be applied only on **Type Declaration emit level** as we consider TypeScript surface as only source of truth for our package APIs.
+
+> This means that if users use only vanilla JavaScript nothing will stop them from using those API's - as we don't support that kind of usage users take the risk that things may/will break for them unfortunately and also they will not get any support from our side.
+
+### Proposed access modifiers:
+
+#### `@internal`
```

Comment:
We could have @public be the modifier and have the tooling make all things @internal by default.  This would force us to be explicit in what is exposed.  If we forget the attribute, the method or type can't get published.

Reply:
I actually agree with this 👍 how easy it is to implement might be a differnt story

Reply:
interesting idea indeed @GeoffCoxMSFT ! that would work only for stable API's though, which is not our use-case as we have various stages of API/packages that are being shipped to users.

> how easy it is to implement might be a differnt story

haha YES